dynamax/nonlinear_gaussian_ssm/inference_ekf.py

```diff
@@ -128,12 +132,28 @@ def _step(carry, t):
         # Predict the next state
         pred_mean, pred_cov = _predict(filtered_mean, filtered_cov, f, F, Q, u)
 
-        return (ll, pred_mean, pred_cov), (filtered_mean, filtered_cov)
+        # Build carry states
+        carry = {
+            "marginal_loglik": ll,
+            "filtered_means": filtered_mean,
+            "filtered_covariances": filtered_cov,
+        }
+        carry = {key: val for key, val in carry.items() if key in outputs}
+
+        return (ll, pred_mean, pred_cov), carry
 
     # Run the extended Kalman filter
     carry = (0.0, params.initial_mean, params.initial_covariance)
-    (ll, _, _), (filtered_means, filtered_covs) = lax.scan(_step, carry, jnp.arange(num_timesteps))
-    return PosteriorGSSMFiltered(marginal_loglik=ll, filtered_means=filtered_means, filtered_covariances=filtered_covs)
```

Comment:
Also confusing since the last outputs are predicted means/covs, not filtered. I think the API would be cleaner if we made the PosteriorGSSMFiltered types optional and then returned Nones for the outputs that are not requested.